### PR TITLE
Mat api cql

### PR DESCRIPTION
**Pull request by jbradl11**
Pull requests into Bonnie require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

Porting over QDM MAT API functionality into Bonnie v2.3 branch. The functionality includes, authentication, user information, and providing a dummy excel export document. 

In order to test, there is a modified MAT API Client branch in gitlab called: "mat_api_cql_client"

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] This PR is into the correct branch.
- [x] JIRA ticket for this PR: https://jira.mitre.org/browse/BONNIE-1528
- [x] JIRA ticket links to this PR
- [x] Code diff has been done and been reviewed (it **does not** contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] If UI changes have been made, google WAVE plug-in has been executed to ensure no 508 issues were introduced.
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass (remember to update Gemfile when applicable)
- [x] Test fixtures updated and documented as necessary ( see [internal wiki](https://gitlab.mitre.org/bonnie/internal-documentation/wikis/testing#test-fixtures) )
- [x] Code coverage has not gone down and all code touched or added is covered. 
     * In rare situations, this may not be possible or applicable to a PR. In those situations:
         1. Note why this could not be done or is not applicable here: A few of the added config files bring down this number below 89.13%. I am open to suggestions. 
         2. Add TODOs in the code noting that it requires a test
         3. Add a JIRA task to add the test and link it here:  No jira test would be valid for this situation.

Branch | Back End Coverage | Front End Coverage
-- | -- | --
bonnie-v2.3 |  1738 / 1950 LOC (89.13%) | image
mat_api_cql | 1878 / 2109 LOC (89.05%) | image

- [x] Automated regression test(s) pass

If JIRA tests were used to supplement or replace automated tests:
- [x] JIRA test links: N/A
- [x] Justification for using JIRA tests: N/A
- [x] JIRA tests have been added to sprint: N/A

**Reviewer 1:**

Name: @hossenlopp
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests: N/A
- [x] JIRA tests have been run and pass N/A
- [x] You agree with the justification for use of JIRA tests or have provided input on why you disagree N/A

**Reviewer 2:**

Name: @mayerm94
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [x] JIRA tests have been run and pass
- [x] You agree with the justification for use of JIRA tests or have provided input on why you disagree

### CODE REVIEW

File: app/controllers/api_v1/measures_controller.rb [hunk 1]
```diff
@@ -101,18 +101,6 @@ def description
     param :id, String, :required => true, :desc => 'The HQMF Set ID of the Measure.'
     error :code => 404, :desc => 'Not Found'
   end
-  
-  def_param_group :measure_upload do
```
Comment (hossenlopp) [inline app/controllers/api_v1/measures_controller.rb]: This shouldn't be removed. It still is useful code for uploading measures which we still need. It does need to be updated for the VSAC profile/release options changes. Perhaps just comment this out for now until we have time to revisit measure uploads.

File: app/controllers/api_v1/measures_controller.rb [hunk 2]
```diff
@@ -223,136 +211,7 @@ def process_patient_records(selector)
     end
     patients
   end
-  
-  def load_measure(params, is_update)
```
Comment (hossenlopp) [inline app/controllers/api_v1/measures_controller.rb]: This is alright to be thrown out. We should re-write this for CQL measures or find a way to make it share code with the regular MeasuresController.

File: config/initializers/webmock.rb [hunk 3]
```diff
@@ -0,0 +1,4 @@
+# Disable WebMock globally so it allows VSAC to work outside of tests
+if defined?(WebMock)
```
Comment (hossenlopp) [inline config/initializers/webmock.rb]: Hmm. I wonder if this is actually needed.

File: config/mongoid.yml [hunk 4]
```diff
@@ -4,6 +4,8 @@ development:
       database: bonnie_development
       hosts:
         - localhost:27017
+  options:
+    raise_not_found_error: true
```
Comment (hossenlopp) [inline config/mongoid.yml]: It still seems odd to me that this is needed...

**Review by hossenlopp — CHANGES_REQUESTED** [review]
A few small changes suggested and comments. There was some code from the MAT API QDM branch that shouldn't be removed, just temporarily disabled/commented out. I hope those comments went through, because I put them on your clean up commits specifically.

File: app/controllers/api_v1/measures_controller.rb (hunk 1 above)
Comment (jbradl11) [reply app/controllers/api_v1/measures_controller.rb]: I am alright with bringing this code back in. I mainly wanted to rip out all unused code, but I am comfortable including this and just commenting it out for now with a TODO.

File: config/mongoid.yml (hunk 4 above)
Comment (jbradl11) [reply config/mongoid.yml]: Good catch. I believe it defaults to `true` even when this is removed. I will remove this.

**Review by hossenlopp — DISMISSED** [review]

**Review by mayerm94 — DISMISSED** [review]

File: config/initializers/doorkeeper.rb [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  #
+  # If not specified, Doorkeeper enables authorization_code and
+  # client_credentials.
+  #
+  # implicit and password grant flows have risks that you should understand
+  # before enabling:
+  #   http://tools.ietf.org/html/rfc6819#section-4.4.2
+  #   http://tools.ietf.org/html/rfc6819#section-4.4.3
+  #
+  grant_flows %w([authorization_code])
```
Comment (hossenlopp) [inline config/initializers/doorkeeper.rb]: This needs to be changed to:
```
 grant_flows %w(authorization_code)
```

**Review by hossenlopp — CHANGES_REQUESTED** [review]

File: app/controllers/api_v1/measures_controller.rb [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    #   param :vsac_date, Date, :required => false, :desc => "The lastest effective date for published value sets to retrieve. Required when include_draft is false."
+    # end
+
+    api :GET, '/api_v1/measures', 'List of Measures'
+    description 'Retrieve the list of measures for the authorized user.'
+    formats [:json, :html]
+    def index
+      # TODO: filter by search parameters, for example an NQF ID or partial description
+      skippable_fields = [:map_fn, :hqmf_document, :oids, :population_ids]
+      @api_v1_measures = CqlMeasure.by_user(current_resource_owner)
```
Comment (hossenlopp) [inline app/controllers/api_v1/measures_controller.rb]: Suggest making the response here smaller by removing `skippable_fields` line and adding to this line to make it get only the fields in MEASURE_WHITELIST defined above in the file.
```
@api_v1_measures = CqlMeasure.by_user(current_resource_owner).only(MEASURE_WHITELIST)
```

Comment (hossenlopp) [inline app/controllers/api_v1/measures_controller.rb]: Suggest making the response here smaller by removing `skippable_fields` line and adding to this line to make it get only the fields in MEASURE_WHITELIST defined above in the file.
```
@api_v1_measures = CqlMeasure.by_user(current_resource_owner).only(MEASURE_WHITELIST)
```

File: app/controllers/api_v1/measures_controller.rb [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+      respond_with @api_v1_measures do |format|
+        format.json do
+          render json: MultiJson.encode(
+            @api_v1_measures.map do |x|
+              h = x
+              h[:id] = x.hqmf_set_id
+              skippable_fields.each { |f| h.delete(f) }
```
Comment (hossenlopp) [inline app/controllers/api_v1/measures_controller.rb]: If using my suggestion above. Remove this line too. Also `.delete` may be destructive to the measure....

**Review by hossenlopp — CHANGES_REQUESTED** [review]
Request making the index response simpler.

**Comment by hossenlopp** [comment]
Give this the sprockets update and I'd say it is good to go. We can handle the application deletion issue later.

**Review by hossenlopp — APPROVED** [review]

**Review by mayerm94 — APPROVED** [review]
Tested all endpoints --they work. Looks good

**Comment by edeyoung** [comment]
checks failing due to sprockets vulnerability. will be updating base branch at a later date. merging.

**Comment by edeyoung** [comment]
@jbradl11 @hossenlopp  can the commit history be cleaned up? Not sure I want to squash merge everything, but there's a lot of extraneous commits.
